src/com/amazon/ion/impl/bin/WriteBuffer.java

```diff
@@ -1272,6 +1272,93 @@ public void writeUInt8At(final long position, final long value)
         block.data[offset] = (byte) value;
     }
 
+    /** Get the length of FlexInt for the provided value. */
+    public static int flexIntLength(final long value) {
+        int numMagnitudeBitsRequired;
+        if (value < 0) {
+            int numLeadingOnes = Long.numberOfLeadingZeros(~value);
+            numMagnitudeBitsRequired = 64 - numLeadingOnes;
+        } else {
+            int numLeadingZeros = Long.numberOfLeadingZeros(value);
+            numMagnitudeBitsRequired = 64 - numLeadingZeros;
+        }
+        return numMagnitudeBitsRequired / 7 + 1;
+    }
+
+    /** Writes a FlexInt to this WriteBuffer, returning the number of bytes that were needed to encode the value */
+    public int writeFlexInt(final long value) {
+        int numBytes = flexIntLength(value);
+        return writeFlexIntOrUInt(value, numBytes);
+    }
+
+    /** Get the length of FlexUInt for the provided value. */
+    public static int flexUIntLength(final long value) {
+        int numLeadingZeros = Long.numberOfLeadingZeros(value);
+        int numMagnitudeBitsRequired = 64 - numLeadingZeros;
+        return (numMagnitudeBitsRequired - 1) / 7 + 1;
+    }
+
+    /** Writes a FlexUInt to this WriteBuffer, returning the number of bytes that were needed to encode the value */
+    public int writeFlexUInt(final long value) {
+        if (value < 0) {
+            throw new IllegalArgumentException("Attempted to write a FlexUInt for " + value);
+        }
+        int numBytes = flexUIntLength(value);
+        return writeFlexIntOrUInt(value, numBytes);
+    }
+
+    /**
+     * Because the flex int and flex uint encodings are so similar, we can use this method to write either one as long
+     * as we provide the correct number of bytes needed to encode the value.
+     */
+    private int writeFlexIntOrUInt(final long value, final int numBytes) {
+        if (numBytes == 1) {
+            writeByte((byte) (0x01 | (byte)(value << 1)));
+        } else if (numBytes == 2) {
+            writeByte((byte) (0x02 | (byte)(value << 2)));
+            writeByte((byte) (value >> 6));
+        } else if (numBytes == 3) {
+            writeByte((byte) (0x04 | (byte)(value << 3)));
+            writeByte((byte) (value >> 5));
+            writeByte((byte) (value >> 13));
+        } else if (numBytes == 4) {
+            writeByte((byte) (0x08 | (byte)(value << 4)));
+            writeByte((byte) (value >> 4));
+            writeByte((byte) (value >> 12));
+            writeByte((byte) (value >> 20));
```

Comment:
As an optimization, we could go into the fallback loop branch if `remaining() < numBytes`, then in the short branches we can write directly into the buffer rather than repetitively calling `writeByte`, which requires some redundant logic. See e.g. `writeUInt64` vs `writeUInt64Slow`.